flocker/node/agents/functional/test_ebs.py

```diff
@@ -156,3 +162,293 @@ def test_next_device_in_use(self):
+        def update(volume):
+            """
+            Transition volume to desired end state and attach data.
+
+            :param boto.ec2.volume.Volume volume: Volume to move to
+                invalid state.
+            """
+            volume.status = self._pick_end_state(operation, state_type)
+            volume.attach_data = self._pick_attach_data(attach_data)
+        return update
+
+    def _assert_fail(self, operation, testcase,
+                     attach_data_type=A.MISSING_ATTACH_DATA):
+        """
+        Assert that configured volume state change for given testcase indicates
+        incomplete operation execution.
+        """
+        volume = self._create_template_ebs_volume(operation)
+        update = self._custom_update(operation, testcase, attach_data_type)
+        start_time = time.time()
+
+        if testcase == self.S.ERROR:
+            self.assertRaises(UnexpectedStateException, _should_finish,
+                              operation, volume, update, start_time, TIMEOUT)
+        else:
+            finish_result = _should_finish(operation, volume, update,
+                                           start_time)
+            self.assertEqual(False, finish_result)
+
+    def _assert_timeout(self, operation, testcase,
+                        attach_data_type=A.MISSING_ATTACH_DATA):
+        """
+        Helper function to validate that ``TimeoutException`` is raised as
+        a result of performing input operation for given testcase on a volume.
+        """
+        volume = self._create_template_ebs_volume(operation)
+        update = self._custom_update(operation, testcase, attach_data_type)
+
+        start_time = time.time()
+        time.sleep(TIMEOUT)
+        self.assertRaises(TimeoutException, _should_finish,
+                          operation, volume, update, start_time, TIMEOUT)
+
+    def _assert_success(self, operation, testcase,
+                        attach_data_type=A.ATTACH_SUCCESS):
+        """
+        Helper function to validate that performing given operation for given
+        testcase on a volume succeeds.
+        """
+        volume = self._create_template_ebs_volume(operation)
+        _wait_for_volume_state_change(operation, volume,
+                                      self._custom_update(operation, testcase,
+                                                          attach_data_type),
+                                      TIMEOUT)
+
+        if operation == self.V.CREATE:
+            self.assertEqual(volume.status, u'available')
+        elif operation == self.V.DESTROY:
+            self.assertEquals(volume.status, u'')
+        elif operation == self.V.ATTACH:
+            self.assertEqual([volume.status, volume.attach_data.device,
+                              volume.attach_data.instance_id],
+                             [u'in-use', u'/dev/sdf', u'i-xyz'])
+        elif operation == self.V.DETACH:
+            self.assertEqual(volume.status, u'available')
```

Comment:
It's generally a good idea to avoid `if` statements in tests (see http://xunitpatterns.com/Conditional%20Test%20Logic.html for excruciating amounts of detail on the subject). I think with `_assert_success` in particular, the tests would be easier to read if there were a helper that creates the volume, waits for state change and then returns the volume. The various tests would then assert things about the returned volume.